Inverse modeling new

xuanluo113 opened the pull request:
Pull request overview
---------------------
Please change this line to a description of the pull request, with useful supporting information including whether it is a new feature, or fixes a defect, a cross reference to any defects addressed in this PR, the type of changes to be made, and whether diffs are expected/justified based on this change.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

Myoldmopar committed:
Add initial proposal document, delete stray design doc

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

Myoldmopar committed:
Functional class structure and successful build with dummy unit test function

Myoldmopar committed:
Starting to stub out functionality

Myoldmopar committed:
Very initial idd object

Myoldmopar committed:
Continued work

Myoldmopar committed:
Starting to extend unit test and factory

Myoldmopar committed:
Continued effort to lay out unit testing

Myoldmopar committed:
Starting to refactor EqFit WWHP code

Myoldmopar committed:
Convert equation fit WWHP model to PlantComponent

Myoldmopar committed:
Fix moved clear_state call

Myoldmopar committed:
Starting to add new eir data; formatted; clean warnings

Myoldmopar committed:
Fix node index problem; add test file

Myoldmopar committed:
Functioning test file

Myoldmopar committed:
Fix unit test, other cleanup

Myoldmopar committed:
Use JSON input structures

Myoldmopar committed:
Clang-format

Myoldmopar committed:
Added some fields to IDD, separated cooling and heating construction, fixed up unit tests

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

Myoldmopar committed:
Added support for pairing up companion heat pump coils, cleaned up unused blocks

Myoldmopar committed:
Start stubbing out support for EIR curve evaluation

Myoldmopar committed:
Handle optional companion coil names, other cleanup

Myoldmopar committed:
Prepping IDF for adding companion coil

Myoldmopar committed:
Break EIR test file into three individual files

Myoldmopar committed:
Fixed sign convention, simulating cooling *and* heating with the simulate() function now

Myoldmopar committed:
Use almost entirely fleshed out EIR implementation

Myoldmopar committed:
Updated, fully functional factory tests

Myoldmopar committed:
Added init tests

Myoldmopar committed:
Applied clang formatting

Myoldmopar committed:
Fix build warning

Myoldmopar committed:
Added test for setpoint worker function -- cooling mode

Myoldmopar committed:
Fix up setpoint unit test; start stubbing out additional tests

Myoldmopar committed:
Clean up tests, move update condenser call, error file has warnings to deal with

Myoldmopar committed:
Fix style in eir header file

Myoldmopar committed:
Start stubbing out simulate() unit test, other minor cleanup

Myoldmopar committed:
Cooling Simulate() unit test fleshed out

Myoldmopar committed:
Added sizing factor input field; hacked together new input processor function for default field value

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

Myoldmopar committed:
Cleanup input processor default function; use custom fixture class

Myoldmopar committed:
Start updating sizing routine

Myoldmopar committed:
Heed @mbadams5 advice on the input processor function

mjwitte committed:
ZoneEquip LoadDist-Add unit test sections to show failures

mjwitte committed:
ZoneEquip LoadDist-DistributeSystemOutputRequired set remaining load to first seq load

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into ZoneEquipLoadDist-6977

jmarrec committed:
Ref #6164: Make Boiler:HotWater Design Water Flow Rate default to Autosize

Also default parasitic electric load to 0

jmarrec committed:
Unit test for #6164

jmarrec committed:
Add unit test for boiler IDD change

jmarrec committed:
Fix #6164: Reset min to max flow rate if min > max and issue warning

jmarrec committed:
Change reset of Min Flow when Min > Max to be 99% of Max, otherwise it doesn't turn on.

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into PR_opened/Fix_6164_BoilerPumpFlowErrors

jmarrec committed:
Update unit test now that reset to 99%

xuanluo113 committed:
NFP

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into inverse-model

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into inverse-model

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into ZoneEquipLoadDist-6977

mjwitte committed:
Fix #7106 - Sequenced loads for unitary with other zone equip

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into ZoneEquipLoadDist-6977

Myoldmopar committed:
Intermediate commit

Myoldmopar committed:
Merge develop; resolve conflict

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into ZoneEquipLoadDist-6977

mjwitte committed:
Initial fix #7110

mjwitte committed:
Plan C to fix zone equipment load distribution with airloops

mjwitte committed:
Zone equipment load distribution - fix build warning

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into ZoneEquipLoadDist-6977

mjwitte committed:
Plan D to fix zone equipment load distribution with airloops

mjwitte committed:
Fix sizing diff in uniform zone equipment load distribution with airloops

mjwitte committed:
Code improvements in zone equipment load distribution

jmarrec committed:
Merge branch 'develop' into PR_opened/Fix_6164_BoilerPumpFlowErrors

tsbyq committed:
Added IM codes

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into inverse-model

tsbyq committed:
Revert "Added IM codes"

This reverts commit cb714468f61e60aaeb57d46308fc433797cf199a.

tsbyq committed:
Resolve conflict with develop

tsbyq committed:
Update Energy+.idd.in

HybridModel:Zone

tsbyq committed:
Added test files for the new HybridModel feature.

tsbyq committed:
Updated IO reference

tsbyq committed:
Update hybrid-model.tex

tsbyq committed:
Added the measurement-unknown parameter relation chart

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

tsbyq committed:
updated the idd

tsbyq committed:
Update DataHeatBalFanSys.cc

Add deallocations

tsbyq committed:
Update ZoneContaminantPredictorCorrector.cc

Decalre varaibles at top of function

tsbyq committed:
Moved the hybrid model output variables to the HybridMode.cc

tsbyq committed:
Add var initialization for NumOccHM

tsbyq committed:
Updated the IO and engineering reference.

tsbyq committed:
Updates for the HybridModel output variables

tsbyq committed:
Added the SumIntGainExceptPeople term in the CalcZoneSums function.

tsbyq committed:
CalcZoneSums overriding removed version.

tsbyq committed:
Applied Clang format.

tsbyq committed:
Added warning and error checks.

Removed the code for solve two or three unknown parameters for now.

tsbyq committed:
Update HybridModel.cc

Added error checks.

Myoldmopar committed:
Merge develop; resolve conflict

Myoldmopar committed:
Fix build issues

tsbyq committed:
Updates to the HybridModel

Update the SetOutputVariable
Update the Calculate Thermal Mass flag
Clean up

tsbyq committed:
Merge branch 'develop' into Inverse_Modeling_new

Sync the upstream to the local new branch - Inverse_Modeling_new.

tsbyq committed:
Updated the IO reference.

tsbyq committed:
Updated the IO references.

- Addressed some of the comments.
- Corrected a typo in the IDD

tsbyq committed:
Merge branch 'develop' into Inverse_Modeling_new

tsbyq committed:
Update format

tsbyq committed:
Remove old example files

xuanluo113 commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -5755,6 +5920,158 @@ namespace ZoneTempPredictorCorrector {
             ZoneAirHumRatTemp(ZoneNum) = RoomAirflowNetworkZoneInfo(ZoneNum).Node(RoomAirflowNetworkZoneInfo(ZoneNum).ControlAirNodeID).HumRat;
         }
 
+        // Han_Li
```
No need to address developer's name.

xuanluo113 commented on src/EnergyPlus/ZoneTempPredictorCorrector.hh:
```diff
@@ -281,6 +281,11 @@ namespace ZoneTempPredictorCorrector {
     );
 
+    // Overload to calculate zone internal heat gains except people
+    void CalcZoneSums(int const ZoneNum,             // Zone number
```
No longer need this declaration?

xuanluo113 commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -5606,14 +5750,15 @@ namespace ZoneTempPredictorCorrector {
-        Real64 MoistureMassFlowRate;
-        Real64 ZoneMassFlowRate;
+        Real64 MoistureMassFlowRate; // Q: Is this the moisture mass flow rate from the system?
+        Real64 ZoneMassFlowRate;     // Q: Same with the one above
```
Does this Q remain unsolved? If solved please delete these comments.

tsbyq committed:
Update group-hybrid-model.tex

xuanluo113 commented on src/EnergyPlus/DataHeatBalance.hh:
```diff
@@ -1171,8 +1171,19 @@ namespace DataHeatBalance {
+		Real64 ZonePeopleSensibleHeatFraction;		  // People activity level input by user
+		Real64 ZonePeopleRadiantHeatFraction;		  // People activity level input by user
+		Real64 ZonePeopleCO2GenerationRate;		  // People activity level input by user
         Real64 ZoneVolCapMultpSens;           // Zone temperature capacity multiplier, i.e. internal thermal mass multiplier
```
Formatting

xuanluo113 commented on src/EnergyPlus/HybridModel.cc:
```diff
@@ -142,35 +165,222 @@ namespace HybridModel {
+                        if (FlagHybridModel_AI) {
+                            ShowContinueError("Field \"" + cAlphaFieldNames(3) + " and " + cAlphaFieldNames(4) + "\" cannot be both set to YES.");
+                            ShowContinueError("Field \"" + cAlphaFieldNames(4) + "\" is changed to NO for the hybrid modeling simulations.");
+                        }
```
These ShowContinueError does not have a ShowSevereError.
ShowSevereError raises an Exception to the program. ShowContinueError just adds up to the logging message.

xuanluo113 replied:
Same as follows.

tsbyq replied:
In this case, when both "calculate thermal mass" and "calculate infiltration" are set to YES, the code should calculate the thermal mass instead of the other. I'm not familiar with the normal practice. Should we terminate the process with the severe error, or continue the simulation but give a error message?

xuanluo113 replied:
@tsbyq If you feel like the program should be terminated, the routine should be
ShowSevereError("blahblah");
ShowContinueError("continue blahblah");
ShowContinueError("continue blahblah");
...

If not (even if the input is not valid, the error is tolerable - such as imported schedule not used), the routine should be:
ShowWarningError("blahblah");
ShowContinueError("continue blahblah");
ShowContinueError("continue blahblah");
...

xuanluo113 commented on src/EnergyPlus/HybridModel.cc:
```diff
@@ -229,7 +464,7 @@ namespace HybridModel {
     {
 
-        FlagHybridModel = false;
+        FlagHybridModel_TM = false;
```
Does any of other state need to be cleared, like FlagHybridModel_AI?

xuanluo113 replied:
For unit test purpose, all changed vars need to be re-initialized. All allocated arrays need to be deallocated.

xuanluo113 commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -7506,6 +7557,30 @@ namespace InternalHeatGains {
     }
 
+    // Added for hybrid model -- function for calculating CO2 gains except people
+    void SumAllInternalCO2GainsExceptPeople(int const ZoneNum, // zone index pointer for which zone to sum gains for
```
I feel like some reviewers' gonna ask why these functions are highly duplicated and code cannot be reused. But we can leave this as a open discussion for now, unless you feel it's easy to modify the code.

xuanluo113 commented on src/EnergyPlus/ZoneContaminantPredictorCorrector.cc:
```diff
@@ -2608,6 +2631,144 @@ namespace ZoneContaminantPredictorCorrector {
                 ZoneAirCO2(ZoneNum) = ZoneAirCO2Temp(ZoneNum);
 
+                // Hybrid modeling with CO2 concentration starts here, added by Han Li 2018
+                GetHybridModelZone();
```
developer & data no longer required

xuanluo113 commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -1821,15 +1822,15 @@ namespace ZoneTempPredictorCorrector {
+                    // multiplier values for the specified zone(s)
                     int ZoneNum = 0;
-                    ZLItem = 0;
+                    ZLItem = 0; // What is this? Used to find the zone number (index) from a list of zones
```
If the comments no longer apply please delete.

xuanluo113 reviewed (commented):
Finished the 2nd round review. Please see the comments on code.

tsbyq committed:
Address comments 1/2

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into ZoneEquipLoadDist-6977

tsbyq committed:
Updated the error check for HybridModel

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into PR_opened/Fix_6164_BoilerPumpFlowErrors

jmarrec committed:
Only issue warning (and compare) if both Min and Nom Volume flow rates aren't autosized

jmarrec committed:
Error in warning message (min above max, not below)

jmarrec committed:
Update Unit Test warning message comparison too

xuanluo113 committed:
first commit

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into inverse-model

xuanluo113 committed:
documentation

xuanluo113 committed:
address CI

tsbyq committed:
Merge branch 'inverse-model' of https://github.com/NREL/EnergyPlus into Inverse_Modeling_new

tsbyq committed:
Update DataHeatBalance.hh

xuanluo113 committed:
CI and example

tsbyq committed:
Start unit testing

tsbyq committed:
Merge branch 'inverse-model' of https://github.com/NREL/EnergyPlus into Inverse_Modeling_new

Myoldmopar committed:
Tidying up before merge

Myoldmopar committed:
Bulk up IDD object

Myoldmopar committed:
Unifying input processors, cleaning up IDD

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

Myoldmopar committed:
Fix unused object properly now

tsbyq committed:
Added a flag, unit test for ZoneTempPredictorCorrector passed.

tsbyq committed:
Removed redundant flags in HybridModel, updated the flag in ZoneTempPredictorCorrector.

Myoldmopar committed:
Fixup unit tests, fix license year

tsbyq committed:
Debug EnergyPlusFixture.ZoneTempPredictorCorrector_CorrectZoneHumRatTest

tsbyq committed:
Updated the unit tests for ZoneContPredicterCorrector

- Updated the HybridModel flag.
- Needs to debug the EnergyPlusFixture.ZoneTempPredictorCorrector_CorrectZoneHumRatTest

Myoldmopar committed:
Added simple coverage result data clear script-needs generalization;
Bulking up test coverage, nearly complete, just a couple topology corner cases and then address sizing - which may be the only functional thing left to do

Myoldmopar committed:
Sizing fully implemented and Bumping up unit test coverage

xuanluo113 committed:
CI

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into inverse-model

xuanluo113 committed:
resolve conflict

Myoldmopar committed:
100% unit test coverage

Myoldmopar committed:
Eliminate unused fields/vars

Myoldmopar committed:
Example file fixed up and running well

Myoldmopar committed:
Remove unused var

tsbyq committed:
Merge branch 'inverse-model' of https://github.com/NREL/EnergyPlus into Inverse_Modeling_new

Myoldmopar committed:
Add IORef

tsbyq committed:
Developing unit test cases for HybriModels

tsbyq committed:
Added unit test cases for HybridModel

Myoldmopar committed:
Add documentation, further unit testing, apply style

xuanluo113 committed:
address comments - 1

Myoldmopar committed:
Wrap up eng ref changes

Myoldmopar committed:
Add snippet to docs to indicate it is constant-flow-request

Myoldmopar committed:
Add constant flow note to idd

Myoldmopar committed:
Fix sign convention issue

tsbyq committed:
Added the test case for solving people count with measured air temperature

tsbyq committed:
Update HybridModel.unit.cc

tsbyq committed:
Update ZoneTempPredictorCorrector.cc

Cleanup

tsbyq committed:
Updated the example files.

tsbyq committed:
Merge branch 'Inverse_modeling_new' into unit_test_develop

tsbyq committed:
Added the test case of solving air infitlration with measured humidity ratio (with HVAC)

Myoldmopar committed:
Merge pull request #6766 from NREL/3982683-EIRFormulatedWWHP

An EIR Formulated Water-to-Water Heat Pump Model

tsbyq committed:
Added the test case of solving people count with measured humidity ratio (with HVAC)

tsbyq committed:
Added the unit test cases for solving infiltration rate and people count with measured CO2 concentration and HVAC supply terms.

Myoldmopar committed:
Merge pull request #7090 from NREL/PR_opened/Fix_6164_BoilerPumpFlowErrors

Improve input error handling in boiler and pump to allow file to run while outputting error messages

tsbyq committed:
Added unit test cases for solving infiltration and people count with measured air temperature and HVAC supply terms.

Commit:
Merge branch 'develop' into ZoneEquipLoadDist-6977

tsbyq committed:
Merge branch 'inverse-model' of https://github.com/NREL/EnergyPlus into Inverse_Modeling_new

tsbyq committed:
Fixed typos

tsbyq committed:
Updates

1. Cleanup
2. Removed unused measurement data.
3. Updated the IDD field notes
4. Added the error message for HybridModel:zone optional fields

tsbyq committed:
Try to separate the inverse modeling logic from the predictor/corrector

tsbyq committed:
Testing the InverseModelingTemperature rotine

tsbyq committed:
sync the HybridModel.unit.cc

tsbyq committed:
Merge branch 'Inverse_Modeling_new' of https://github.com/tsbyq/EnergyPlus into Inverse_Modeling_new

tsbyq committed:
New InverseModelTemperature passed unit tests.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into inverse-model

mjwitte committed:
HybridModel - fix formatting errors - removing tabs

mjwitte committed:
HybridModel - transition rules

mjwitte committed:
HybridModel - transition code

mjwitte committed:
HybridModel - fix unit test failures

xuanluo113 committed:
conflict

xuanluo113 committed:
Merge branch 'Inverse_Modeling_new' of https://github.com/tsbyq/EnergyPlus into Inverse_Modeling_new

Myoldmopar committed:
Merge pull request #7025 from NREL/ZoneEquipLoadDist-6977

Fix Zone Equipment Load Distribution

tsbyq committed:
Update ZoneContaminantPredictorCorrector.unit.cc

Fixe the unit test issue.

tsbyq committed:
New InverseModelTemperature subroutine passed unit tests.

xuanluo113 committed:
fix tests

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into inverse-model

tsbyq committed:
New InverseModelCO2 subroutine passed unit tests.

xuanluo113 committed:
resolve conflict

tsbyq committed:
Merge remote-tracking branch 'upstream/inverse-model' into Inverse_Modeling_new

tsbyq committed:
Delete old test files.

tsbyq committed:
Merge branch 'new_HM_logic' into Inverse_Modeling_new

tsbyq committed:
Fixed merge issues.

tsbyq committed:
Update group-hybrid-model.tex

Addressed the formatting and field name reference comments.

tsbyq committed:
Updated the IDD and I/O reference

Added clarification for default and NO cases.

tsbyq committed:
Updated the HybridModel zone flags.

tsbyq committed:
Uncomment one unit test case.
